### PR TITLE
add Search Trait

### PR DESCRIPTION
**Pull request by at-duongtruong**
## Description
Add Search Trait
## Reference Requirements
- Product Backlog ID: PBI-6
- Task ID: #6.2
## Migrations
YES

## Impacted Areas in Application
List general components of the application that this PR will affect:

-

## SQL query and `explain`
select * from `users` where `id` = '11' and `users`.`deleted_at` is null limit 1

select count(*) as aggregate from `books` left join `borrows` on `books`.`id` = `borrows`.`id` where ((`title` LIKE '%Dr%' or `author` LIKE '%Dr%')) and `books`.`deleted_at` is null group by `books`.`id`

select `books`.`id`, `books`.`title`, `books`.`author`, `books`.`rating`, COUNT(borrows.id) AS total_borrowed from `books` left join `borrows` on `books`.`id` = `borrows`.`id` where ((`title` LIKE '%Dr%' or `author` LIKE '%Dr%')) and `books`.`deleted_at` is null group by `books`.`id` order by `id` desc limit 10 offset 0

select count(*) as aggregate from `books` left join `borrows` on `books`.`id` = `borrows`.`id` where `title` LIKE '%Dr%' and `books`.`deleted_at` is null group by `books`.`id`

select `books`.`id`, `books`.`title`, `books`.`author`, `books`.`rating`, COUNT(borrows.id) AS total_borrowed from `books` left join `borrows` on `books`.`id` = `borrows`.`id` where `title` LIKE '%Dr%' and `books`.`deleted_at` is null group by `books`.`id` order by `id` desc limit 10 offset 0

select count(*) as aggregate from `books` left join `borrows` on `books`.`id` = `borrows`.`id` where `author` LIKE '%Dr%' and `books`.`deleted_at` is null group by `books`.`id`

select `books`.`id`, `books`.`title`, `books`.`author`, `books`.`rating`, COUNT(borrows.id) AS total_borrowed from `books` left join `borrows` on `books`.`id` = `borrows`.`id` where `author` LIKE '%Dr%' and `books`.`deleted_at` is null group by `books`.`id` order by `id` desc limit 10 offset 0

### CODE REVIEW

File: app/Model/Book.php [hunk 1]
```diff
@@ -121,4 +122,19 @@ protected static function boot()
...
+    protected $searchableFields = [
+        'columns' => [
+            'title',
+            'author',
+        ],
+        'joins' => [
+            'borrows' => ['books.id', 'borrows.id']
+        ]
+    ];
 }
```
Comment (at-trampham) [inline app/Model/Book.php]: LGTM

File: app/Http/Controllers/Admin/BookController.php [hunk 2]
```diff
@@ -25,7 +24,6 @@ public function index(Request $request)
     {
         $search = $request->search;
         $filter = $request->filter;
```
Comment (at-tuanphan) [inline app/Http/Controllers/Admin/BookController.php]: $search = $request->search;
 $filter = $request->filter;
không cần phải gán biến đâu em, chổ nào dùng thì request trở đến luôn

File: app/Http/Controllers/Admin/BookController.php [hunk 3]
```diff
@@ -133,52 +120,14 @@ public function create()
...
-            return redirect()->route('books.index');
-        } catch (\Exception $e) {
-            if (isset($pictureName) && \File::exists($folderStore.$pictureName)) {
-                \File::delete($folderStore.$pictureName);
-            }
-            DB::rollback();
-            flash(__('books.create_failure'))->error();
-            return redirect()->back()->withInput();
-        }
+        $title = $request->title;
```
Comment (at-tuanphan) [inline app/Http/Controllers/Admin/BookController.php]: vì sao lại xoá luôn cái hàm tạo book rồi em?

File: app/Model/Book.php [hunk 4]
```diff
@@ -121,4 +122,19 @@ protected static function boot()
...
     }
+
+     /**
+     * The attributes that can be search.
+     *
+     * @var array $searchableFields
+     */
+    protected $searchableFields = [
+        'columns' => [
+            'title',
```
Comment (at-tuanphan) [inline app/Model/Book.php]: dùng những biến mà mình đã define ở bên trên đi em:
self::TYPE_TITLE áp dụng cho những cái khác nữa nhé

**Review by ghost — COMMENTED** [review]
please check my comments

File: app/Http/Controllers/Admin/BookController.php [hunk 5]
```diff
@@ -26,9 +25,8 @@ class BookController extends Controller
      */
     public function index(Request $request)
     {
-        $search = $request->search;
-        $filter = $request->filter;
-
+        // $search = $request->search;
```
Comment (at-tuanphan) [inline app/Http/Controllers/Admin/BookController.php]: không dùng thì xoá luôn đi chứ em ơi 👯‍♂️

File: app/Http/Controllers/Admin/BookController.php [hunk 6]
```diff
@@ -169,7 +156,7 @@ public function create()
     /**
      * Store a newly created resource in storage.
      *
-     * @param CreateBookRequest $request send request
+     * @param CategoryRequest $request send request
```
Comment (at-tuanphan) [inline app/Http/Controllers/Admin/BookController.php]: hàm tạo sách sao lại là `CategoryRequest`

**Review by ghost — COMMENTED** [review]
please check my comments

**Review by ghost — APPROVED** [review]
LGTM

**Pull request merged**